File: 733-flood-fill/733-flood-fill.js
/**
 * @param {number[][]} image
 * @param {number} sr
 * @param {number} sc
 * @param {number} color
 * @return {number[][]}
 */
var floodFill = function(image, sr, sc, color) {
    //const to have original value
    const original = image[sr][sc];
    
    //function recursion to check all side
    function floodFill(image, sr, sc){
        //if out of bounds or not the original value to fill or already have color then just return
        if(sr < 0 || sr > image.length -1 || sc < 0 || sc > image[0].length -1 || image[sr][sc] !== original || image[sr][sc] === color){
            return image;
        }
        
        //color the image if nothing above happens
        image[sr][sc] = color;
        
        //now recursion call to all 4 directions
        floodFill(image, sr+1, sc);
        floodFill(image, sr-1, sc);
        floodFill(image, sr, sc+1);
        floodFill(image, sr, sc-1);
        
        return image;
    }
    
    return floodFill(image, sr, sc);
    
};

File: 733-flood-fill/733-flood-fill.java
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        int original = image[sr][sc];
        
        return floodFillRecursion(image, sr, sc, original, color);
    }
    
    //recursino call to check for 4 directions
    public int[][] floodFillRecursion(int[][] image, int sr, int sc, int original, int color){
        //base cse of out of bounds or already colored or not the original to return image
        if(sr < 0 || sr > image.length -1 || sc < 0 || sc >image[0].length-1 || image[sr][sc] != original || image[sr][sc] == color){
            return image;
        }
        
        //color the block and check for all directions
        image[sr][sc] = color;
        floodFillRecursion(image, sr+1, sc, original, color);
        floodFillRecursion(image, sr-1, sc, original, color);
        floodFillRecursion(image, sr, sc+1, original, color);
        floodFillRecursion(image, sr, sc-1, original, color);
        return image;
    }
}